paywall/src/unlock.js/Wallet.ts

```diff
@@ -0,0 +1,315 @@
+
+      this.iframes.data.postMessage(PostMessages.SEND_UPDATES, 'locks')
+    })
+
+    // listen for account and network from the user accounts iframe
+    this.iframes.accounts.on(PostMessages.UPDATE_ACCOUNT, account => {
+      this.userAccountAddress = account
+    })
+    this.iframes.accounts.on(PostMessages.UPDATE_NETWORK, network => {
+      this.userAccountNetwork = network
+    })
+
+    // when a purchase is in progress, tell the data iframe to retrieve the transaction
+    this.iframes.accounts.on(PostMessages.INITIATED_TRANSACTION, () => {
+      this.iframes.data.postMessage(
+        PostMessages.INITIATED_TRANSACTION,
+        undefined
+      )
+    })
+
+    // then create the iframe and ready its post office
+    this.iframes.accounts.createIframe()
+  }
+
+  setupProxyWallet() {
+    if (this.useUserAccounts && !this.hasWallet) {
+      // if user accounts are explicitly enabled, we use them
+      // but only if there is no crypto wallet
+      this.setupUserAccountsProxyWallet()
+    } else {
+      // if we have a wallet, we always use it
+      // if we have no wallet, and no use accounts, we use the web3 proxy wallet
+      this.setupWeb3ProxyWallet()
+    }
+  }
+
+  async enableCryptoWallet() {
+    if (
+      !this.window.web3 ||
+      !this.window.web3.currentProvider ||
+      !this.window.web3.currentProvider.enable
+    ) {
+      return
+    }
+    await this.window.web3.currentProvider.enable()
+  }
+
+  /**
+   * This is the proxy wallet for a crypto wallet
+   *
+   * It handles the "ready" event posted by the Web3ProxyProvider
+   * in the data iframe, and also the "web3" event for
+   * communicating with the crypto wallet
+   */
+  setupWeb3ProxyWallet() {
+    // when receiving a key purchase request, we either pass it to the
+    // account iframe for credit card purchase if user accounts are
+    // explicitly enabled, or to the crypto wallet
+    this.iframes.checkout.on(PostMessages.PURCHASE_KEY, async request => {
+      this.iframes.data.postMessage(PostMessages.PURCHASE_KEY, request)
+    })
+
+    this.iframes.data.on(PostMessages.READY_WEB3, async () => {
+      // initialize, we do this once the iframe is ready to receive information on the wallet
+      // we need to tell the iframe if the wallet is metamask
+      // TODO: pass the name of the wallet if we know it? (secondary importance right now, so omitting)
+      if (!this.hasWallet) {
+        // the user has no crypto wallet
+        this.hasWeb3 = false
+        this.iframes.data.postMessage(PostMessages.WALLET_INFO, {
+          noWallet: true,
+          notEnabled: false,
+          isMetamask: false,
+        })
+        return
+      }
+      try {
+        // first, enable the wallet if necessary
+        await this.enableCryptoWallet()
+      } catch (e) {
+        // user declined to enable the crypto wallet
+        // they still have a wallet, but we need to re-enable it to use it
+        this.hasWeb3 = true
```

Comment:
Move that line above the `try/catch` so that you do not have to duplicate it?

Reply:
good call. done.